### DISCUSSION TITLE
OpenType features with private Unicode code points

### DISCUSSION THREAD

**Discussion by IvanUkhov** [harfbuzz/harfbuzz#4805] [Q&A] [answered]
Hello,

This is more of a question. I am experimenting with Proxima Nova's features, and I see that `ordn`, Ordinals, does not work as it should in Chrome (Blink), while it works as expected in Safari (WebKit). More specifically, there is set of substitutions making _th_ jump into the superscript whenever the _th_ is preceded by a ordinal numeral ending in _th_ in English, like fourth. It is realized as follows:

```
sub [zero four five six seven eight nine] t' h by uniF6F3;
sub [zero four five six seven eight nine] uniF6F3 h' by hsuperior;
```

`uniF6F3` is a glyph representing a superscript _t_ mapped to U+F6F3, and so it is accessible directly. (On the other hand, `hsuperior` is not mapped to any Unicode code point and hence indirectly accessible.)

So the issue is that Blink does not process the second rule on its own: if I type `4` + U+F6F3 + `h`, for example, it does not replace the _h_ with its subscript counterpart, while WebKit does, as illustrated below.

I wonder if it has something to do with U+F6F3 residing in the plane for private use. And is it a deliberate choice, or is it a bug?

macOS Chrome:

<img width="668" alt="Skärmavbild 2024-07-16 kl  10 38 13" src="https://github.com/user-attachments/assets/5c465ee5-4566-499c-b7d8-4aff7707f7e4">

iOS Chrome:

<img width="383" alt="Skärmavbild 2024-07-16 kl  10 38 54" src="https://github.com/user-attachments/assets/3a76c313-e0ba-40b0-a4fe-5b0fc3907ea0">

**Comment by khaledhosny** [comment, marked as answer]
Since you are inputting Private Use Area code points directly, this is likely a text segmentation issue. Chrome is probably not grouping PUA code points with the surrounding characters because PUA has Zzzz (uncoded) script. Try Firefox, which should be treating Zzzz similar to Zyyy (undetermined) script and groups it with the surrounding characters. Safari/CoreText might be doing something similar.

But I’m not sure why you are inputting U+F6F3 instead of `t`, which shouldn’t trigger such text segmentation issue.

Anyway, text segmentation is done before HarfBuzz is called, so if there is an issue here, it would be a Chrome issue.

**Reply by IvanUkhov** [reply]
I see. Thank you for the prompt answer!

> But I’m not sure why you are inputting U+F6F3 instead of t, which shouldn’t trigger such text segmentation issue.

It is for visualization of features. You extract what is in the font and show it all. I will ask why that glyph is mapped that way. It could have easily stayed as an internal glyph.
